include/circt/Dialect/ESI/cosim/Server.h

```diff
@@ -0,0 +1,90 @@
+//===- Server.h - ESI cosim RPC servers -------------------------*- C++ -*-===//
+//
+// Various classes used to implement the RPC server classes generated by
+// CapnProto. Capnp C++ RPC servers are based on 'libkj' and its asyncrony
+// model, which is very foreign.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_DIALECT_ESI_COSIM_SERVER_H
+#define CIRCT_DIALECT_ESI_COSIM_SERVER_H
+
+#include "circt/Dialect/ESI/cosim/CosimDpi.capnp.h"
+#include "circt/Dialect/ESI/cosim/Endpoint.h"
+
+namespace std {
+class thread;
+} // namespace std
+
+namespace circt {
+namespace esi {
+namespace cosim {
+
+/// Implements the `EsiDpiEndpoint` interface from the RPC schema. Mostly a
+/// wrapper around an `Endpoint` object. Whereas the `Endpoint`s are long-lived
+/// (associated with the RTL endpoint), this class is constructed/destructed
+/// when the client open()s it.
+class EndpointServer final
+    : public EsiDpiEndpoint<capnp::AnyPointer, capnp::AnyPointer>::Server {
+  /// The wrapped endpoint.
+  Endpoint &endpoint;
+  /// Signals that this endpoint has been opened by a client and hasn't been
+  /// closed by said client.
+  bool open;
+
+public:
+  EndpointServer(Endpoint &ep);
+  /// Release the Endpoint should the client disconnect without properly closing
+  /// it.
+  ~EndpointServer();
+  /// Disallow copying as the 'open' variable needs to track the endpoint.
+  EndpointServer(const EndpointServer &) = delete;
+
+  /// Implement the EsiDpiEndpoint RPC interface.
+  kj::Promise<void> send(SendContext);
```

Comment:
What is kj::Promise?  Where did this come from?

Reply:
It's from `libkj`, the C++ library upon which the capnproto C++ implementation depends. It contains an asynchronous framework and (as you see below) an alternative collections library among other stuff. The generated (by capnpc) C++ code expects to have the RPC server implemented by methods like this.

I've added a note to the header comment.

Reply:
Aha, I see.  Is it possible/reasonable to move these references to the .cpp files and out of the public headers?

Three ways to go: move some of the public headers into the lib directory (so they are private headers), use the pImpl idiom to hide some of these things within a library, or use forward declarations for these if they don't actually need to be exposed to library clients.

Reply:
It is! Turns out the code which needs it is contained to one file.

Reply:
Oh man that feels so much better. We keep all the awkward capnp and libkj code in ONE file! (And in an anonymous namespace!) Thanks for pushing on this @lattner.